src/kmod/file-write-tracking.c

```diff
@@ -0,0 +1,367 @@
+					  entry->last_tgid, GFP_KERNEL);
+	if (!event) {
+		hashtbl_free_generic(file_write_table, entry);
+		return false;
+	}
+
+	event->fileGeneric.file_type = entry->state.fileType;
+	memcpy(event->fileGeneric.path, entry->path.buf, PATH_MAX);
+	logger_submit_event(event);
+	ret = true;
+
+	hashtbl_free_generic(file_write_table, entry);
+
+	return ret;
+}
+
+bool set_file_entry_data(struct file *file, pid_t *last_tgid,
+			 struct file_type_state *state, const char *path)
+{
+	bool			 found;
+	struct hashtbl_bkt *	 bkt   = NULL;
+	unsigned long		 flags = 0;
+	struct file_write_entry *entry = NULL;
+	struct file_write_key	 key   = { .file = file };
+	if (!file || (!path && !state && !last_tgid)) {
+		return false;
+	}
+
+	// lock bucket
+	found = hashtbl_getlocked_bucket(file_write_table, &key,
+					 (void **)&entry, &bkt, &flags);
+	if (found) {
+		if (entry) {
+			if (last_tgid) {
+				entry->last_tgid = *last_tgid;
+			}
+			if (state) {
+				memcpy(&entry->state, state, sizeof(*state));
+			}
+			if (path) {
+				strncpy(entry->path.buf, path, PATH_MAX);
+			}
+		}
+		if (bkt) {
+			hashtbl_unlock_bucket(bkt, flags);
+		}
+	}
+
+	return found;
+}
+
+bool get_file_entry_data(struct file *file, pid_t *last_tgid,
+			 struct file_type_state *state,
+			 struct file_write_path *path)
+{
+	bool			 found;
+	struct hashtbl_bkt *	 bkt   = NULL;
+	unsigned long		 flags = 0;
+	struct file_write_entry *entry = NULL;
+	struct file_write_key	 key   = { .file = file };
+
+	if (!file || (!state && !path && !last_tgid)) {
+		return false;
+	}
+
+	// lock bucket
+	found = hashtbl_getlocked_bucket(file_write_table, &key,
+					 (void **)&entry, &bkt, &flags);
+	if (found) {
+		if (entry) {
+			if (last_tgid) {
+				*last_tgid = entry->last_tgid;
+			}
+			if (state) {
+				memcpy(state, &entry->state, sizeof(*state));
+			}
+			if (path) {
+				memcpy(path, &entry->path, sizeof(*path));
+			}
+		}
+		if (bkt) {
+			hashtbl_unlock_bucket(bkt, flags);
+		}
+	}
+	return found;
+}
+
+static char *getTypeStr(enum CB_FILE_TYPE type)
+{
+	char *str = "unknown";
+
+	switch (type) {
+	case filetypePe:
+		str = "PE";
+		break;
+	case filetypeElf:
+		str = "ELF";
+		break;
+	case filetypeUniversalBin:
+		str = "Univ. Bin";
+		break;
+	case filetypeEicar:
+		str = "EICAR";
+		break;
+	case filetypeOfficeLegacy:
+		str = "Office Legacy";
+		break;
+	case filetypeOfficeOpenXml:
+		str = "Office Open XML";
+		break;
+	case filetypePdf:
+		str = "PDF";
+		break;
+	case filetypeArchivePkzip:
+		str = "PKZIP";
+		break;
+	case filetypeArchiveLzh:
+		str = "LZH";
+		break;
+	case filetypeArchiveLzw:
+		str = "LZW";
+		break;
+	case filetypeArchiveRar:
+		str = "RAR";
+		break;
+	case filetypeArchiveTar:
+		str = "TAR";
+		break;
+	case filetypeArchive7zip:
+		str = "7 ZIP";
+		break;
+	case filetypeUnknown:
+	default:
+		break;
+	}
+	return str;
+}
+
+static int _show_file_tracking_table(struct HashTbl *	   hashTblp,
```

Comment:
Remove unused parameter `struct HashTbl * hashTblp`

Reply:
Pretty sure it's a required parameter for a function pointer